source/simulator/stokes_matrix_free.cc

```diff
@@ -1598,6 +1677,128 @@ namespace aspect
+                                         "compute derivatives."));
+
+                      const double viscosity_derivative_wrt_pressure
+                        = derivatives->viscosity_derivative_wrt_pressure[q];
+
+                      const SymmetricTensor<2,dim> viscosity_derivative_wrt_strain_rate
+                        = derivatives->viscosity_derivative_wrt_strain_rate[q];
+
+                      // use the spd factor when the stabilization is PD or SPD
+                      const double alpha =  (sim.newton_handler->parameters.velocity_block_stabilization
+                                             & Newton::Parameters::Stabilization::PD)
+                                            != Newton::Parameters::Stabilization::none
+                                            ?
+                                            Utilities::compute_spd_factor<dim>(out.viscosities[q], strain_rate, viscosity_derivative_wrt_strain_rate,
+                                                                               sim.newton_handler->parameters.SPD_safety_factor)
+                                            :
+                                            1;
+
+                      active_viscosity_derivative_wrt_pressure_table(cell,q)[i]
+                        = viscosity_derivative_wrt_pressure * newton_derivative_scaling_factor;
+
+                      for (unsigned int m=0; m<dim; ++m)
+                        for (unsigned int n=0; n<dim; ++n)
+                          {
+                            active_strain_rate_table(cell, q)[m][n][i]
+                              = strain_rate[m][n];
+
+                            active_viscosity_derivative_wrt_strain_rate_table(cell, q)[m][n][i]
+                              = derivatives->viscosity_derivative_wrt_strain_rate[q][m][n]
+                                * newton_derivative_scaling_factor * alpha;
+                          }
+                    }
+                }
+            }
+          const bool is_symmetric =
+            ((sim.newton_handler->parameters.preconditioner_stabilization
+              & Newton::Parameters::Stabilization::symmetric)
+             != Newton::Parameters::Stabilization::none);
+
+          stokes_matrix.fill_Newton_cell_data(active_viscosity_derivative_wrt_pressure_table,
+                                              active_strain_rate_table,
+                                              active_viscosity_derivative_wrt_strain_rate_table,
+                                              is_symmetric);
+        }
+      else
+        {
+          active_strain_rate_table.reinit(TableIndices<2>(0,0));
+          active_viscosity_derivative_wrt_pressure_table.reinit(TableIndices<2>(0,0));
+          active_viscosity_derivative_wrt_strain_rate_table.reinit(TableIndices<2>(0,0));
+
+          stokes_matrix.fill_Newton_cell_data(active_viscosity_derivative_wrt_pressure_table,
```

Comment:
passing these options like this is not very elegant, but I would propose to merge as is and I prepare a followup.